src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,202 @@
+                        ? (Action<TypeImportCompletionItemInfo>)HandlePublicAndInternalItem
+                        : HandlePublicItem;
+
+            // Add only public types to completion list
+            void HandlePublicItem(TypeImportCompletionItemInfo itemInfo)
+                => AddItems(itemInfo, isInternalsVisible: false, completionContext, namespacesInScope);
+
+            // Add both public and internal types to completion list
+            void HandlePublicAndInternalItem(TypeImportCompletionItemInfo itemInfo)
+                => AddItems(itemInfo, isInternalsVisible: true, completionContext, namespacesInScope);
+
+            static void AddItems(TypeImportCompletionItemInfo itemInfo, bool isInternalsVisible, CompletionContext completionContext, HashSet<string> namespacesInScope)
+            {
+                if (itemInfo.IsPublic || isInternalsVisible)
+                {
+                    var containingNamespace = TypeImportCompletionItem.GetContainingNamespace(itemInfo.Item);
+                    if (!namespacesInScope.Contains(containingNamespace))
+                    {
+                        completionContext.AddItem(itemInfo.Item);
+                    }
+                }
+            }
+        }
+
+        private HashSet<string> GetNamespacesInScope(Document document, SyntaxContext syntaxContext, CancellationToken cancellationToken)
+        {
+            var semanticModel = syntaxContext.SemanticModel;
+            var importedNamespaces = GetImportedNamespaces(syntaxContext.LeftToken.Parent, semanticModel, cancellationToken);
+
+            // This hashset will be used to match namespace names, so it must have the same case-sensitivity as the source language.
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var namespacesInScope = new HashSet<string>(importedNamespaces, syntaxFacts.StringComparer);
+
+            // Get containing namespaces.
+            var namespaceSymbol = semanticModel.GetEnclosingNamespace(syntaxContext.Position, cancellationToken);
+            while (namespaceSymbol != null)
+            {
+                namespacesInScope.Add(namespaceSymbol.ToDisplayString(SymbolDisplayFormats.NameFormat));
+                namespaceSymbol = namespaceSymbol.ContainingNamespace;
+            }
+
+            return namespacesInScope;
+        }
+
+        public override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey = default, CancellationToken cancellationToken = default)
+        {
+            var newDocument = await ComputeNewDocumentAsync(document, item, cancellationToken).ConfigureAwait(false);
+            var newText = await newDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+            var changes = await newDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
+            var change = Utilities.Collapse(newText, changes.ToImmutableArray());
```

Comment:
>So if you're in a huge file towards the bottom and we invoke this, we're collapsing the edit down into a single edit that'll have a "NewText" of the entire file? Oof. 

Note exactly.  The roslyn-data layer exposes it this way.  However, the application to the buffer still does it as an edit that will be made with minimal changes.

thsi was done because the original code that had to deal with many text-changes was just enormously buggy.  We could try to bring it back, but we'd just need to be super delicate about how it was done this time around.